### PR TITLE
Patched svcCreateThread() for core 2 and 3 on new 3DS

### PR DESCRIPTION
**Pull request by Core-2-Extreme**
Patch native.firm to always be able to create thread on core 2 and 3 from user app.
I just patched `if(processorid == 0x2) to if(processorid == 0x4)` and `if(processorid == 0x3) to if(processorid == 0x4)` so it won't return 0xD9001BEA.

I tested it and worked on new 3DS firm version 11.15.0.47J, 11.15.0.47U and 11.14.0.46J.

### CODE REVIEW

**Review by Frenzoid — APPROVED** [review]

**Comment by PabloMK7** [comment]
The patches made to kernel11 are hardcoded, so it will only ever work with a single firmware version (which you don't specify btw :P) The way to fix this would be to use pattern search like all the other patches.

**Comment by Core-2-Extreme** [comment]
I updated to pattern search.

File: arm9/source/patches.c [hunk 1]
```diff
@@ -709,3 +709,49 @@ u32 patchAgbBootSplash(u8 *pos, u32 size)
 
     return 0;
 }
+
+u32 patchCore2(u8 *pos, u32 size)
+{
+    //cmp r9, #0x2
+    //beq #0x38
+    static const u8 pattern[] = { 0x02, 0x00, 0x59, 0xE3, 0x0C, 0x00, 0x00, 0x0A, };
```
Comment (PabloMK7) [inline arm9/source/patches.c]:
```suggestion
    static const u8 pattern[] = { 0x02, 0x00, 0x59, 0xE3, 0x0C, 0x00, 0x00, 0x0A };
```

File: arm9/source/patches.c [hunk 2]
```diff
@@ -709,3 +709,49 @@ u32 patchAgbBootSplash(u8 *pos, u32 size)
...
+        return 1;
+
+    /*
+    Patch svcCreateThread()
+    if(processorid == 0x2 && memory_region != BASE && !(kernel_flags & 0x2000)) return 0xD9001BEA;
+    -> if(processorid == 0x4 && memory_region != BASE && !(kernel_flags & 0x2000)) return 0xD9001BEA;
+    So we can always create thread on core #2 even process is not running on BASE memory region.
+    */
+    //cmp r9, #0x2 -> cmp r9, #0x4
+    offset[1] = __builtin_bswap32(0x040059E3);
```
Comment (PabloMK7) [inline arm9/source/patches.c]:
```suggestion
    offset[1] = 0xE3590004
```

File: arm9/source/patches.c [hunk 3]
```diff
@@ -709,3 +709,49 @@ u32 patchAgbBootSplash(u8 *pos, u32 size)
...
+
+    return 0;
+}
+
+u32 patchCore3(u8 *pos, u32 size)
+{
+    //cmp r9, #0x3
+    //beq #0x3C
+    static const u8 pattern[] = { 0x03, 0x00, 0x59, 0xE3, 0x0D, 0x00, 0x00, 0x0A, };
```
Comment (PabloMK7) [inline arm9/source/patches.c]:
```suggestion
    static const u8 pattern[] = { 0x03, 0x00, 0x59, 0xE3, 0x0D, 0x00, 0x00, 0x0A };
```

File: arm9/source/patches.c [hunk 4]
```diff
@@ -709,3 +709,49 @@ u32 patchAgbBootSplash(u8 *pos, u32 size)
...
+        return 1;
+
+    /*
+    Patch svcCreateThread()
+    if(processorid == 0x3 && memory_region != BASE && !(kernel_flags & 0x2000)) return 0xD9001BEA;
+    -> if(processorid == 0x4 && memory_region != BASE && !(kernel_flags & 0x2000)) return 0xD9001BEA;
+    So we can always create thread on core #3 even process is not running on BASE memory region.
+    */
+    //cmp r9, #0x3 -> cmp r9, #0x4
+    offset[1] = __builtin_bswap32(0x040059E3);
```
Comment (PabloMK7) [inline arm9/source/patches.c]:
```suggestion
    offset[1] = 0xE3590004
```

**Review by PabloMK7 — CHANGES_REQUESTED** [review]
Check suggested changes. Also, your patterns seem to have branch instructions in them. Are you sure they are valid for all firmware versions?

**Comment by Core-2-Extreme** [comment]
I'm not sure, and I don't have old `native.firm` so I couldn't check it.

**Comment by PabloMK7** [comment]
You didn't apply the suggestions, but marked them as resolved, was that intended? lol

**Comment by Core-2-Extreme** [comment]
No, I thought if I press the resolve button, it'll be merged in my code.
I think I applied your suggestions now, if it is wrong, let me know.

**Comment by TuxSH** [comment]
Is there a use case for the core2 patch?

Core2 access is granted via a kernel capability, which we automatically grant to all 3dsx apps.

**Comment by PabloMK7** [comment]
> Is there a use case for the core2 patch?

For the same reason as svc and service patches, I guess :P. [Also 3GX plugins](https://gitlab.com/thepixellizeross/ctrpluginframework/-/blob/f7aed655468af724ab78bb347bf5ece7369a3ccc/Library/source/ctrulibExtension/thread.c#L81-83).

**Comment by Core-2-Extreme** [comment]
> Is there a use case for the core2 patch?

Does luma3ds grant access to core 2 even if kernel-flags bitmask 0x2000 is not set?
> When processorid==0x2 and the process is not a BASE mem-region process, exheader kernel-flags bitmask 0x2000 must be set (otherwise error 0xD9001BEA is returned).
(https://www.3dbrew.org/wiki/Multi-threading)

I was thinking that if homebrew launcher (any app that starts `.3dsx` app) doesn't have it then `.3dsx` app can't get access to core 2 results in some app not being worked in `.3dsx` while it works in `.cia`.

**Comment by TuxSH** [comment]
> Does luma3ds grant access to core 2 even if kernel-flags bitmask 0x2000 is not set?

Yes, it sets it when creating a process.

The homebrew launcher app's exheader or code is never actually used, Luma3DS hijacks the loading of the target program (you can set homebrew loader titleId to match the Download Play app, for example).

> For the same reason as svc and service patches

Those haven't been "patches" for quite a long time.

**Comment by Core-2-Extreme** [comment]
> Yes, it sets it when creating a process.

ok, then core2 patch is not necessary, but should I remove it?
(I think it won't hurt anything)

**Comment by PabloMK7** [comment]
I see this as a positive addition, I don't think it hurts having it.

**Comment by TuxSH** [comment]
Sorry if I took a long time to deal with this. I ended up writing the patch differently.

For what it's worth https://github.com/devkitPro/3ds-examples/blob/master/threads/thread-basic/source/main.c modified to use core 3 for all spawned threads often resulted in one or multiple threads being starved in my tests.

**Comment by Core-2-Extreme** [comment]
> For what it's worth https://github.com/devkitPro/3ds-examples/blob/master/threads/thread-basic/source/main.c modified to use core 3 for all spawned threads often resulted in one or multiple threads being starved in my tests.

Did you fix this problem in your new patch?
If not, is there different between my patch?

**Comment by TuxSH** [comment]
No, please read the commit message. This is not something you can fix, QTM is running at very high priority on core3, and the GPU driver depends on QTM on N3DS.

User applications are not meant to use core3 at all. Just use core2 instead.

**Pull request closed without merging**